### DISCUSSION TITLE
How to run PHP code from within a Go application?

### DISCUSSION THREAD

**Discussion by nickchomey** [php/frankenphp#1022] [Q&A]
My application is primarily written in PHP, but I am starting to add some background services written in Go. Is there a way to use Frankenphp's innovations to execute existing PHP code/scripts from within the Go application? 

Is it with the [ExecuteScriptCLI function](https://pkg.go.dev/github.com/dunglas/frankenphp#ExecuteScriptCLI)? Can this essentially run an entire php application if the script you call bootstraps itself with require/include?

Thanks!

**Comment by withinboredom** [comment]
Hey hey,

I'm using this feature on an experimental runtime I've been working on for the last couple of years. It's nowhere near finished or even production ready; but you can take a look at the code: https://github.com/bottledcode/durable-php/blob/v2/cli/glue/glue.go#L147-L209

I went with creating a custom "request" and using a [custom ResponseWriter](https://github.com/bottledcode/durable-php/blob/v2/cli/glue/response_writer.go) to intercept and log output.

`ExecuteScript` works pretty well, but you can't do much with the output.

At work, we use this same method of custom requests to read from a queue (in go), transform it to a request, and have php process the job. When we return to go, the output may contain other messages to send. It's like 300 lines of code in go, and there would be thousands of lines in php with lots of headaches to manage processes, memory, etc. Sure, there are opinionated php libraries to make it easier, but you can't beat 300 lines of non-opinionated code.

**Reply by nickchomey** [reply]
Thanks very much!

At a glance, it looks like our interests very much overlap. A big portion of what I'm using go for is NATS, which is absolute magic.

Do you mind if I ask some more questions about what you're doing, after I've had time to explore the code more closely?

**Reply by withinboredom** [reply]
v3 (which isn't public yet) is much better than v2 and uses an actual wire protocol over an `os.pipe` to communicate to worker threads; v2 uses files in /tmp which isn't great, but worked for v2. Happy to answer any question you might have though.
